CesiumUtility/src/Uri.cpp

```diff
@@ -200,6 +201,131 @@ std::string Uri::escape(const std::string& s) {
+  // it.
+  std::string windowsPathClean;
+  windowsPathClean.resize(windowsPath.size());
+  std::replace_copy(
+      windowsPath.begin(),
+      windowsPath.end(),
+      windowsPathClean.begin(),
+      '/',
+      '\\');
+
+  // UriParser docs:
+  //   The destination buffer must be large enough to hold 8 + 3 * len(filename)
+  //   + 1 characters in case of an absolute filename or 3 * len(filename) + 1
+  //   in case of a relative filename.
+  std::string result(8 + 3 * windowsPathClean.size() + 1, '\0');
+
+  if (uriWindowsFilenameToUriStringA(windowsPathClean.data(), result.data()) !=
+      0) {
+    // Error - return original string.
+    return windowsPath;
+  } else {
+    // An absolute URI will start with "file://". Remove this.
+    if (result.find("file://", 0, 7) != std::string::npos) {
+      result.erase(0, 7);
+    }
+
+    // Truncate at first null character
+    result.resize(std::strlen(result.data()));
+    return result;
+  }
+}
+
+std::string Uri::nativePathToUriPath(const std::string& nativePath) {
+  if constexpr (std::filesystem::path::preferred_separator == '\\') {
+    return windowsPathToUriPath(nativePath);
+  } else {
+    return unixPathToUriPath(nativePath);
+  }
+}
+
+std::string Uri::uriPathToUnixPath(const std::string& uriPath) {
+  // UriParser docs:
+  //   The destination buffer must be large enough to hold len(uriString) + 1
+  //   - 5 characters in case of an absolute URI or len(uriString) + 1 in case
+  //   of a relative URI.
+  // However, the above seems to assume that uriPath starts with "file:", which
+  // is not required.
+  std::string result(uriPath.size() + 1, '\0');
+  if (uriUriStringToUnixFilenameA(uriPath.data(), result.data()) != 0) {
+    // Error - return original string.
+    return uriPath;
+  } else {
+    // Truncate at first null character
+    result.resize(std::strlen(result.data()));
+    return result;
+  }
+}
+
+std::string Uri::uriPathToWindowsPath(const std::string& uriPath) {
+  // If the URI starts with `/c:` or similar, remove the initial slash.
+  size_t skip = 0;
+  if (uriPath.size() >= 3 && uriPath[0] == '/' && uriPath[1] != '/' &&
+      uriPath[2] == ':') {
+    skip = 1;
+  }
+
+  // UriParser docs:
+  //   The destination buffer must be large enough to hold len(uriString) + 1
+  //   - 5 characters in case of an absolute URI or len(uriString) + 1 in case
+  //   of a relative URI.
+  // However, the above seems to assume that uriPath starts with "file:", which
+  // is not required.
+  std::string result(uriPath.size() + 1, '\0');
+  if (uriUriStringToWindowsFilenameA(uriPath.data() + skip, result.data()) !=
+      0) {
+    // Error - return original string.
+    return uriPath;
+  } else {
+    // Truncate at first null character
+    result.resize(std::strlen(result.data()));
+    return result;
+  }
+}
+
+std::string Uri::uriPathToNativePath(const std::string& nativePath) {
+  if constexpr (std::filesystem::path::preferred_separator == '\\') {
```

Comment:
If the goal is to detect whether we are on windows or not, can we just use `#ifdef _WIN32`?

Reply:
I had that first, but I think this is better. The main difference between the unix and windows functions in UriParser is their handling of backslashes, so I think choosing based on the platform path separator is slightly better.

Reply:
Ok and you've well documented the behavior, so all good.